packages/apps/job-launcher/server/src/modules/job/routing-protocol.service.ts

```diff
@@ -1,34 +1,142 @@
+  private readonly reputationOraclePriorityOrder: number[];
+  private chainCurrentIndex = 0;
+  private reputationOracleIndex = 0;
+  private oraclesHash: string | null = null;
+  private oracleOrder: OracleOrder = {};
 
-  constructor() {
+  constructor(
+    private readonly web3Service: Web3Service,
+    private readonly web3ConfigService: Web3ConfigService,
+  ) {
     this.chains = Object.keys(NETWORKS).map((chainId) => +chainId);
-    this.currentIndex = 0;
-    this.priorityOrder = this.generatePriorityOrder();
+    this.reputationOracles = this.web3ConfigService.reputationOracles
+      .split(',')
+      .map((address) => address.trim());
+
+    this.chainPriorityOrder = this.shuffleArray(this.chains);
+    this.reputationOraclePriorityOrder = this.shuffleArray(
+      this.reputationOracles.map((_, i) => i),
+    );
+
+    this.oracleOrder = this.chains.reduce((acc: OracleOrder, chainId) => {
+      acc[chainId] = this.reputationOracles.reduce(
+        (oracleAcc, reputationOracle) => {
+          oracleAcc[reputationOracle] = {
+            [Role.ExchangeOracle]: [],
+            [Role.RecordingOracle]: [],
+          };
+          return oracleAcc;
+        },
+        {} as { [reputationOracle: string]: { [role: string]: string[] } },
+      );
+      return acc;
+    }, {} as OracleOrder);
   }
 
-  private generatePriorityOrder() {
-    const priorityOrder = Array.from(Array(this.chains.length).keys());
-    for (let i = priorityOrder.length - 1; i > 0; i--) {
-      const j = Math.floor(Math.random() * (i + 1));
-      [priorityOrder[i], priorityOrder[j]] = [
-        priorityOrder[j],
-        priorityOrder[i],
-      ];
-    }
-    return priorityOrder;
+  private shuffleArray<T>(array: T[]): T[] {
+    return array.sort(() => Math.random() - 0.5);
   }
 
   public selectNetwork(): ChainId {
-    const chainId = this.chains[this.priorityOrder[this.currentIndex]];
-    this.currentIndex = (this.currentIndex + 1) % this.chains.length;
+    const chainId =
+      this.chains[this.chainPriorityOrder[this.chainCurrentIndex]];
+    this.chainCurrentIndex = (this.chainCurrentIndex + 1) % this.chains.length;
     return chainId;
   }
+
+  public selectReputationOracle(): string {
+    const reputationOracle =
+      this.reputationOracles[
+        this.reputationOraclePriorityOrder[this.reputationOracleIndex]
+      ];
+
+    this.reputationOracleIndex =
+      (this.reputationOracleIndex + 1) % this.reputationOracles.length;
+    return reputationOracle;
+  }
+
+  public async selectOracleFromAvailable(
+    availableOracles: any[],
+    oracleType: string,
+    chainId: ChainId,
+    reputationOracle: string,
+  ): Promise<string | null> {
+    const oraclesOfType = availableOracles
+      .filter((oracle) => oracle.role === oracleType)
+      .map((oracle) => oracle.address);
+
+    if (!oraclesOfType.length) {
+      return null;
+    }
+
+    const latestOraclesHash = hashString(JSON.stringify(oraclesOfType));
```

Comment:
The `selectOracleFromAvailable` function is designed to select an appropriate oracle of a specified type (e.g., exchange oracle or recording oracle) from a set of available oracles for a specific chain and reputation oracle.

Parameters:
* availableOracles: An array of oracles that are available for selection. Each oracle has attributes like role (defining the type of the oracle) and address (the oracle’s address).
* oracleType: A string representing the type of oracle to select (e.g., Role.ExchangeOracle, Role.RecordingOracle).
* chainId: The ID of the blockchain network.
* reputationOracle: The address of the reputation oracle to be used for the oracle selection.

The `oraclesHash` is used to track whether the list of available oracles has changed. It stores the hash of the most recent list of oracles. When the function is called, it compares the hash of the current list of oracles with oraclesHash:
* If the hash has changed (i.e., the list of available oracles is different), the list is shuffled, and the order is stored in oracleOrder.
* If the hash remains the same, it reuses the previously shuffled order.

Reply:
As you can see, the `selectOracleFromAvailable` function is called only from the `selectOracles` function and takes in a list of available oracles selected from the subgraph, in order to avoid accessing the subgraph several times when selecting oracles.

Reply:
`oraclesHash` is used for optimization, since the data in the subgraph can change, and the service can find out about this only upon receiving updated data, during the launch of the service, the routing protocol creates an order in which oracles will be selected, if the data changes during runtime, then the routing service will find out about this and create a new order based on the updated data.